### PR TITLE
save/restore tab input mode

### PR DESCRIPTION
**Pull request by mjauvin**
new pull request based on master this time.

when switching tabs, save current tab input mode and restore new tab input mode

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3481)

### CODE REVIEW

File: qutebrowser/browser/browsertab.py [hunk 1]
```diff
@@ -650,6 +650,7 @@ def __init__(self, *, win_id, mode_manager, private, parent=None):
         # self.action = AbstractAction()
 
         self.data = TabData()
+        self._input_mode = usertypes.KeyMode.normal
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py]: This is a private attribute, so you shouldn't read/write it from the outside. This should probably go into the `TabData` object instead.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 2]
```diff
@@ -659,7 +662,8 @@ def on_current_changed(self, idx):
                                usertypes.KeyMode.passthrough]
         for mode in modes_to_leave:
             modeman.leave(self._win_id, mode, 'tab changed', maybe=True)
-
+            
+        modeman.enter(self._win_id, tab._input_mode, 'restore input mode for tab')
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: There are multiple issues with this:

- If the old tab was e.g. in hint/caret/prompt/... mode, you'll end up in that mode again, but without the prompt/hints/etc. leading to undefined behaviour or crashes. You probably only want to re-enter insert/passthrough mode, no?
- It breaks the `tabs.persist_mode_on_change` setting.
- It should probably have its own setting as well. Maybe that should even be a merged `tabs.mode_on_change` setting or so. I can take care of that part (after my exams) if you prefer.

Comment (mjauvin) [reply qutebrowser/mainwindow/tabbedbrowser.py]: I have addressed all the noted issues and will provide patches for these soon. 

Thanks for the nice and insightful feedback!

**Comment by The-Compiler** [comment]
You seem to keep accidentally pushing unrelated changes :wink:

For your next PR, you might want to look into [creating a branch](https://guides.github.com/introduction/flow/) (`git checkout -b branchname` when you're on the latest `master`) instead.

**Comment by mjauvin** [comment]
@The-Compiler  I would like to close this PR in order to resubmit from a branch in my fork since it looks like I cannot submit new independent PR because of this

Is it fine with you?

**Comment by The-Compiler** [comment]
Go ahead! If you need help, I'd suggest [joining our IRC chat](https://github.com/qutebrowser/qutebrowser#getting-help), it's probably easier to figure things out there than it is here on GitHub :smile:

**Comment by mjauvin** [comment]
Closing this and restart from a reserved branch

**Pull request closed without merging**